src/adapter/src/explain_new/optimizer_trace.rs

```diff
@@ -0,0 +1,128 @@
+use mz_ore::tracing::{PlanTrace, TraceEntry};
+use mz_repr::{
+    explain_new::{Explain, ExplainConfig, ExplainError, ExplainFormat},
+    Timestamp,
+};
+use mz_sql::plan::HirRelationExpr;
+use tracing::dispatcher::{self, with_default};
+use tracing_subscriber::prelude::*;
+
+use crate::{catalog::ConnCatalog, coord::peek::FastPathPlan};
+
+use super::{ExplainContext, Explainable, UsedIndexes};
+
+type LirRelationExpr = Plan<Timestamp>;
+
+/// Provides functionality for tracing plans generated by the execution of an
+/// optimization pipeline.
+pub(crate) struct OptimizerTrace(dispatcher::Dispatch);
+
+impl OptimizerTrace {
+    pub fn new() -> OptimizerTrace {
+        let subscriber = tracing_subscriber::registry()
+            .with(PlanTrace::<HirRelationExpr>::new())
+            .with(PlanTrace::<MirRelationExpr>::new())
+            .with(PlanTrace::<DataflowDescription<OptimizedMirRelationExpr>>::new())
+            .with(PlanTrace::<DataflowDescription<LirRelationExpr>>::new());
+
+        OptimizerTrace(dispatcher::Dispatch::new(subscriber))
+    }
+
+    /// Run the given optimization `pipeline` once and collect a trace of all
+    /// plans produced during that run.
+    pub fn collect_trace<T>(&self, pipeline: impl FnOnce() -> T) -> T {
+        with_default(&self.0, pipeline)
+    }
+
+    /// Collect all traced plans for all plan types `T` that are available in
+    /// the wrapped [`dispatcher::Dispatch`].
+    pub fn drain_all<'a>(
+        self,
+        format: ExplainFormat,
+        config: ExplainConfig,
+        catalog: ConnCatalog<'a>,
+        row_set_finishing: Option<RowSetFinishing>,
+        used_indexes: Vec<mz_repr::GlobalId>,
+        fast_path_plan: Option<FastPathPlan<Timestamp>>,
+    ) -> Result<Vec<TraceEntry<String>>, ExplainError> {
+        let mut results = vec![];
+
+        // drain plans from local stages
+        let context = ExplainContext {
+            config: &config,
+            humanizer: &catalog,
+            used_indexes: UsedIndexes::new(vec![]),
+            finishing: row_set_finishing.clone(),
+            fast_path_plan: None,
+        };
+        results.extend(itertools::chain!(
+            self.drain_all_of::<HirRelationExpr>(&format, &context)?,
+            self.drain_all_of::<MirRelationExpr>(&format, &context)?,
+        ));
+
+        // drain plans from global stages
+        let context = ExplainContext {
+            config: &config,
+            humanizer: &catalog,
+            used_indexes: UsedIndexes::new(used_indexes),
+            finishing: row_set_finishing,
+            fast_path_plan,
+        };
+        results.extend(itertools::chain!(
+            self.drain_all_of::<DataflowDescription<OptimizedMirRelationExpr>>(&format, &context)?,
+            self.drain_all_of::<DataflowDescription<LirRelationExpr>>(&format, &context)?,
+        ));
+
+        // sort plans by instant
+        results.sort_by(|x, y| x.instant.cmp(&y.instant));
```

Comment:
I think `sort_by_key` can be used here and would be more efficient:

```
results.sort_by_key(|x| x.instant);
```

... or `sort_unstable_by_key`, if unstable sorting is fine.